### DISCUSSION TITLE
Preserving client side state, specifically `details`

### DISCUSSION THREAD

**Discussion by yatesco** [bigskysoftware/htmx#1004] [Q&A] [answered]
Hi - how should I go about preserving client side state between updates. Specifically whether a `details` element has been opened or not?

I can POST the data to the server but the `details` element is literally right next to a `textarea` whose `mouseout` event reiggers a POST/refresh, so there is a non-trivial likelihood that the user will change the open/close before the response comes back, which would be nasty.

Is there a more `htmx`/`hyperscript` way than simply capturing it via JS and then doing something on the relevant `htmx` refreshed event?

Thanks!

**Comment by David-Guillot** [comment]
Hi @yatesco , have you tried  [hx-preserve](https://htmx.org/attributes/hx-preserve/) ?

**Comment by yatesco** [comment]
Thanks David, yes I looked at that, but I do actually want to replace the ‘details’ element.

**Comment by David-Guillot** [comment, marked as answer]
Ah yes... In this case it's a bit trickier. I have one more idea for you, but I haven't tested it. It would be to combine [hx-preserve](https://htmx.org/attributes/hx-preserve/) and [hx-swap-oob](https://htmx.org/attributes/hx-swap-oob/)  to:

* Avoid replacing the `<details>`, in order to preserve its state
* But replace its contents as if it was out-of-band target

On server side, you'd have to:

* Wrap you details contents in some `<div>` which would hold the id that's required for `hx-swap-oob`
* Put the details contents at the root of the HTML fragment returned in case of htmx request

I hope this is clear enough...

**Reply by yatesco** [reply]
thanks David - I might give that a go.

**Comment by eduardofcgo** [comment]
I had a similar problem, my first solution was just as @yatesco described, but I later found it unnecessary complex.

 I stopped using the details and used my own custom element. The details contents will be inside this element as well, but I created also an adicional element to hold the collapsed state. This element is the one tagged with `hx-preserve`. Because the preserved element is different from the actual contents, you can swap freely without disturbing the UI "state". Of course, the collapsed effect can be archived with a sibling CSS combinator rule according to this preserved state element.

**Comment by norbertkeri** [comment]
I ended up here after searching for "hx-preserve not working properly with oob" on Kagi, and I can't seem to make this work in 2.0.2. I think the bug described below is affecting this:

https://github.com/bigskysoftware/htmx/issues/2922

In short, `hx-preserve` is currently not taken into account for oob swaps.
